cub/cub/detail/triple_chevron_launch.cuh

```diff
@@ -0,0 +1,154 @@
+ * Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of the NVIDIA CORPORATION nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+#pragma once
+
+#include <thrust/detail/config.h>
+#include <thrust/system/cuda/detail/core/alignment.h>
+#include <thrust/system/cuda/detail/guarded_cuda_runtime_api.h>
+
+#include <cuda/std/version> // _LIBCUDACXX_HIDDEN
+
+#include <cassert>
+
+CUB_NAMESPACE_BEGIN
+
+
+namespace detail
+{
+
+  struct _LIBCUDACXX_HIDDEN triple_chevron
+  {
+    typedef size_t Size;
+    dim3 const grid;
+    dim3 const block;
+    Size const shared_mem;
+    cudaStream_t const stream;
+
+    CUB_RUNTIME_FUNCTION // Should this be CUB_RUNTIME_FUNCTION now?
+    triple_chevron(dim3         grid_,
+                   dim3         block_,
+                   Size         shared_mem_ = 0,
+                   cudaStream_t stream_     = 0)
+        : grid(grid_),
+          block(block_),
+          shared_mem(shared_mem_),
+          stream(stream_) {}
+
+    template<class K, class... Args>
+    cudaError_t __host__
+    doit_host(K k, Args const&... args) const
+    {
+      k<<<grid, block, shared_mem, stream>>>(args...);
+      return cudaPeekAtLastError();
+    }
+
+    template<class T>
+    size_t __device__
+    align_up(size_t offset) const
+    {
+      size_t alignment = THRUST_NS_QUALIFIER::cuda_cub::alignment_of<T>::value;
+      return alignment * ((offset + (alignment - 1))/ alignment);
```

Comment:
Is this way of accessing the namespace ok? Since `alignment_of` exists in thrust(and is seemingly not used at all through out CUB), does it make sense to continue relying on it? I was curious to know if it would be better to decouple the two. Thrust makes use of `alignment_of` more. Just wondering if it's best to decouple as much as possible from Thrust for CUB.

Reply:
That way is indeed necessary to avoid any ambiguity. Note that the macro has a leading :: in the namespace which ensures that we do not catch relative namespaces